### PR TITLE
ZIP-321 Parsing Feature

### LINKED ISSUE #30: Add Parsing support
Support parsing URI Strings into data structures

### PR DESCRIPTION
**Pull request by pacu**
Closes #30 

Closes https://github.com/pacu/zwcd/issues/74

### CODE REVIEW

File: Sources/zcash-swift-payment-uri/Model.swift [hunk 1]
```diff
@@ -141,7 +150,23 @@ public struct MemoBytes {
         
         self.data = memoStringData
     }
-    
+
+    public init(base64URL: String) throws {
+        var base64 = base64URL.replacingOccurrences(of: "_", with: "/")
+            .replacingOccurrences(of: "-", with: "+")
```
Comment (nuttycom) [inline Sources/zcash-swift-payment-uri/Model.swift:156]: What are these replacements for?

File: Sources/zcash-swift-payment-uri/Parser.swift [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+//
+//  Created by Francisco Gindre on 2023-13-08.
+//
+
+import Foundation
+import Parsing
+
+/// Result of the URI parser. This returns a `PaymentRequest` if a compliant
+/// [ZIP-321](https://zips.z.cash/zip-0321) request is parsed from
+/// a URI String. This implementation accounts of legacy schemes like that preceded
```
[the remark this answers is not in the document]
Comment (pacu) [reply Sources/zcash-swift-payment-uri/Parser.swift:13]: thanks for clarifying this point @nuttycom.

I guess we need to rework this section 

> Backward compatibility 
As this ZIP is written, several clients already implement a zcash: URI scheme similar to this one, however usually without the additional req- prefix requirement or the facility to specify multiple payments using paramindex. These implementations also generally do not support URIs, even with a single payment, where the address is specified as an address= query parameter rather than in the hier-part of the URI. They may also not support the memo parameter, or may not treat it as base64url-encoded.


because then it's unclear to me as reader what should I do with this contextual information. If the answer is "probably nothing" then maybe we should rephrase this to be more straight forward about that.

[the remark this answers is not in the document]
Comment (pacu) [reply Sources/zcash-swift-payment-uri/Parser.swift:13]: opened issue about this https://github.com/pacu/zcash-swift-payment-uri/issues/34

File: Sources/zcash-swift-payment-uri/Model.swift (hunk 1 above)
Comment (pacu) [reply Sources/zcash-swift-payment-uri/Model.swift:156]: So the `Data(base64Encoded: base64)` does not fail
References: 
https://stackoverflow.com/questions/43499651/decode-base64url-to-base64-swift
https://en.wikipedia.org/wiki/Base64#Variants_summary_table
https://en.wikipedia.org/wiki/Base64#URL_applications

File: Sources/zcash-swift-payment-uri/Parser.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (nuttycom) [reply Sources/zcash-swift-payment-uri/Parser.swift:13]: The intention of this is to state that this ZIP is intended to be backwards-compatible with such usage, that's all.
